### PR TITLE
Added support for label_size=1 as an optional parameter for csv.

### PR DESCRIPTION
**Pull request by aws-patlin**
*Issue #, if available:*
https://issues.amazon.com/issues/AWSMLC-77

*Description of changes:*
Added additional handling in data_utils.py to check for 'text/csv' content type with the ``label_size`` optional parameter. If the ``label_size`` is present and is set to 1, the content type is accepted, otherwise a ``UserError`` is raised.

*Testing:*
Added two assert statements in the ``test_get_content_type`` unit test. Tests were run with tox and passed successfully.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-patlin** [comment]
Is it not a concern that user may include other optional parameters, such as 'text/csv; charset=utf-8'? If this happens, I think it makes sense for us to allow this, even if it is not officially supported.

**Comment by ericangelokim** [comment]
I'm not saying we shouldn't do it, but we should wait to have the larger discussion with support for this across all containers.

**Comment by aws-patlin** [comment]
Okay, in that case, is this bare-bones fix the right approach for now?

**Comment by edwardjkim** [comment]
Do we allow `text/csv;label_size=1` (with no spaces after the semicolon)?. This [RCF](https://docs.aws.amazon.com/sagemaker/latest/dg/randomcutforest.html) documentation has `text/csv;label_size=1` and `text/csv;label_size=0 ` as examples. How about two (or more) spaces?

**Comment by aws-patlin** [comment]
Yury notes that we should be as generally permissive of optional parameters and formatting. Reverted to a more general style of checking for optional parameters.

File: src/sagemaker_xgboost_container/data_utils.py [hunk 1]
```diff
@@ -60,6 +60,16 @@ def get_content_type(content_type_cfg_val):
         return LIBSVM
     elif content_type_cfg_val.lower() in [CSV, _content_types.CSV]:
         return CSV
+    elif _content_types.CSV in content_type_cfg_val.lower():
+        items = content_type_cfg_val.split(';')
+        label_size = [item for item in items if 'label_size' in item]
```
Comment (iyerr3) [inline src/sagemaker_xgboost_container/data_utils.py]: I would suggest renaming the variable to something similar to `all_label_sizes` to indicate that this is an iterable.

File: src/sagemaker_xgboost_container/data_utils.py [hunk 2]
```diff
@@ -60,6 +60,16 @@ def get_content_type(content_type_cfg_val):
...
+        if label_size:
+            label_size = label_size[0].split('=')[1]
```
Comment (iyerr3) [inline src/sagemaker_xgboost_container/data_utils.py]: Why is this only for the first label size? If we're making a generic solution, then might as well make it `any(label_size != 1 ... )`

File: test/unit/test_data_utils.py [hunk 3]
```diff
@@ -31,9 +31,14 @@ def test_get_content_type(self):
...
         self.assertEqual('csv', data_utils.get_content_type('csv'))
         self.assertEqual('csv', data_utils.get_content_type('text/csv'))
+        self.assertEqual('csv', data_utils.get_content_type('text/csv; label_size=1'))
+        self.assertEqual('csv', data_utils.get_content_type('text/csv;label_size = 1'))
+        self.assertEqual('csv', data_utils.get_content_type('text/csv; charset=utf-8'))
 
         with self.assertRaises(exc.UserError):
             data_utils.get_content_type('incorrect_format')
+        with self.assertRaises(exc.UserError):
+            data_utils.get_content_type('text/csv; label_size=5')
```
Comment (iyerr3) [inline test/unit/test_data_utils.py]: how about `text/csv; label_size`, `text/csv; label_size=1=1` and `text/csv; label_size=1; label_size=2`?

Comment (aws-patlin) [reply test/unit/test_data_utils.py]: Good point, I wasn't thinking of these kinds of edge cases.

File: src/sagemaker_xgboost_container/data_utils.py [hunk 4]
```diff
@@ -60,6 +60,16 @@ def get_content_type(content_type_cfg_val):
         return LIBSVM
     elif content_type_cfg_val.lower() in [CSV, _content_types.CSV]:
         return CSV
+    elif _content_types.CSV in content_type_cfg_val.lower():
```
Comment (rizwangilani) [inline src/sagemaker_xgboost_container/data_utils.py]: From a readability perspective, I find this a little confusing as there is an explicit csv condition right above this (on line 61).
Would it make more sense to handle this in a single if condition (i.e. if its a csv) and work on label_size from there?

File: src/sagemaker_xgboost_container/data_utils.py [hunk 5]
```diff
@@ -60,6 +60,16 @@ def get_content_type(content_type_cfg_val):
...
+        items = content_type_cfg_val.split(';')
+        label_size = [item for item in items if 'label_size' in item]
+        if label_size:
+            label_size = label_size[0].split('=')[1]
+            if label_size.strip() != '1':
```
Comment (rizwangilani) [inline src/sagemaker_xgboost_container/data_utils.py]: How about making 1 as a parameter?

File: src/sagemaker_xgboost_container/data_utils.py [hunk 6]
```diff
@@ -60,6 +60,16 @@ def get_content_type(content_type_cfg_val):
...
+                msg = "{} is not an accepted csv ContentType. "\
+                      "Optional parameter label_size must be equal to 1".format(content_type_cfg_val)
```
Comment (rizwangilani) [inline src/sagemaker_xgboost_container/data_utils.py]: can reference the same parameter here

File: src/sagemaker_xgboost_container/data_utils.py (hunk 4 above)
Comment (iyerr3) [reply src/sagemaker_xgboost_container/data_utils.py]: Good point - I think the lines 61-62 need to be removed.

File: src/sagemaker_xgboost_container/data_utils.py (hunk 1 above)
Comment (rizwangilani) [reply src/sagemaker_xgboost_container/data_utils.py]: do we expect more than 1 label_size? is that a valid case?

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: I feel like that should raise an exception, if multiple label_size values are given.

File: src/sagemaker_xgboost_container/data_utils.py (hunk 5 above)
Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: What would be the benefit in this case? By definition, label size for XGBoost must be 1, and this value is only referenced twice.

File: src/sagemaker_xgboost_container/data_utils.py [hunk 7]
```diff
@@ -43,13 +43,49 @@ def _get_invalid_csv_error_msg(line_snippet, file_name):
     return INVALID_CONTENT_FORMAT_ERROR.format(line_snippet=line_snippet, file_name=file_name, content_type='CSV')
 
 
+def _get_csv_content_type(content_type_cfg_val):
+    """
+    Returns CSV if content_type_cfg_val is
```
Comment (ericangelokim) [inline src/sagemaker_xgboost_container/data_utils.py]: Minor: PEP8, make this a command rather than desc ie ```Return CSV if content_type ...```

See examples here: https://www.python.org/dev/peps/pep-0257/

File: src/sagemaker_xgboost_container/data_utils.py [hunk 8]
```diff
@@ -25,8 +25,8 @@
 LIBSVM = 'libsvm'
 
 
-INVALID_CONTENT_TYPE_ERROR = "{invalid_content_type} is not an accepted ContentType:" \
-                             " 'csv', 'libsvm', 'text/csv', 'text/libsvm', 'text/x-libsvm'."
+INVALID_CONTENT_TYPE_ERROR = "{invalid_content_type} is not an accepted ContentType: " \
```
Comment (ericangelokim) [inline src/sagemaker_xgboost_container/data_utils.py]: Minor. I would've expected ```text/csv;label_size=1``` without space. Did you check that this is the conventional pattern?

File: src/sagemaker_xgboost_container/data_utils.py [hunk 9]
```diff
@@ -43,13 +43,49 @@ def _get_invalid_csv_error_msg(line_snippet, file_name):
...
+    """
+    Returns CSV if content_type_cfg_val is
+    * 'csv',
+    * 'text/csv',
+    * 'text/csv; ...' with valid parameters
+    """
+    if content_type_cfg_val in [CSV, _content_types.CSV]:
+        # Allow 'csv' and 'text/csv'
+        return CSV
+    elif _content_types.CSV in content_type_cfg_val:
```
Comment (ericangelokim) [inline src/sagemaker_xgboost_container/data_utils.py]: This seems like a really ugly way to deal with a general problem. Have you looked into existing tools that can parse request header content types? such as: https://pypi.org/project/python-mimeparse/

It just feels like we are trying to reinvent the wheel here.

File: src/sagemaker_xgboost_container/data_utils.py (hunk 8 above)
Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: Yes, the public sagemaker documentation shows a space: https://docs.aws.amazon.com/sagemaker/latest/dg/cdf-training.html

File: src/sagemaker_xgboost_container/data_utils.py (hunk 9 above)
Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: Good point, I hadn't looked into existing tools.

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: Looks like python has a library called 'cgi' with a [parse_header()](https://docs.python.org/2/library/cgi.html#cgi.parse_header) function which we can leverage. My only concern is that it handles duplicate parameters by keeping the last one (i.e. 'text/csv; label_size=1; label_size=2' produces a parameter dictionary with {'label_size': '2'}). Is this acceptable behavior?

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: It also returns an empty dictionary if the parameter is only a name without a value (i.e. 'text/csv; label_size' -> `('text/csv', {})`), which would pass instead of erroring.

Comment (iyerr3) [reply src/sagemaker_xgboost_container/data_utils.py]: Does the standard for MIME/HTTP headers mention rules for options in content-type? 
If not then IMO it's better to be permissive than restrictive for optional parameters.

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: The protocol doesn't have any guidelines regarding these edge cases: https://www.w3.org/Protocols/rfc1341/4_Content-Type.html
I'd say it should be fine to just go with the default behavior provided by cgi.parse_header()

File: src/sagemaker_xgboost_container/data_utils.py [hunk 10]
```diff
@@ -25,8 +26,8 @@
 LIBSVM = 'libsvm'
 
 
-INVALID_CONTENT_TYPE_ERROR = "{invalid_content_type} is not an accepted ContentType:" \
-                             " 'csv', 'libsvm', 'text/csv', 'text/libsvm', 'text/x-libsvm'."
```
Comment (ericangelokim) [inline src/sagemaker_xgboost_container/data_utils.py]: Minor: These accepted ContentTypes should be put in a list so that we aren't copy/pasting raw string values. No need to submit a review for this change.

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: Do you mean just for this error message, or for all references to these content types? I went ahead and created a list just for the error message, but it might be worthwhile to migrate everything into a dictionary (though this would be a separate issue to deal with).

Comment (ericangelokim) [reply src/sagemaker_xgboost_container/data_utils.py]: There should be a list that can be referenced throughout the codebase so that we aren't checking against a copy pasted list repeatedly. This isn't just for the error messages but throughout the repo.

Comment (aws-patlin) [reply src/sagemaker_xgboost_container/data_utils.py]: That would mean changing the way we're referencing these content types on a large scale. We should definitely do this, but I'm not sure this PR is the right place to make this change. What do you think?

Comment (ericangelokim) [reply src/sagemaker_xgboost_container/data_utils.py]: That's fair. Can we make the list for the error messages for now, since they refer to the same exact list? No need to resubmit for approval.
